### PR TITLE
Remote updates

### PR DESCRIPTION
**Pull request by clockard**
#### What does this PR do?
Adds the ability to sync remote changes of replicated data back to the origin. There is no merging of changes just simple latest modified time wins

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@shaundmorris 
@sgalla

#### What are the relevant tickets?
Fixes NA-72

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: adapters/ddf-adapter/src/main/java/com/connexta/replication/adapters/ddf/csw/CswMinimalRecordConverter.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      LOGGER.debug("Unable to convert date {} from default locale format {} ", value, e);
+    }
+
+    // default to current date
+    LOGGER.debug("Unable to convert {} to a date object", value);
+    return null;
+  }
+
+  public static Metadata createMetadataFromCswRecord(HierarchicalStreamReader reader) {
+    String id = reader.getAttribute("id");
```
Comment (shaundmorris) [inline adapters/ddf-adapter/src/main/java/com/connexta/replication/adapters/ddf/csw/CswMinimalRecordConverter.java:87]: This shouldn't be null, but do we need to check for that?

File: replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/ReplicationItemManagerImpl.java [hunk 2]
```diff
@@ -54,15 +58,22 @@
...
 
   @Override
   public Optional<ReplicationItem> getItem(String metadataId, String source, String destination) {
+    if (cache.containsKey(metadataId + source + destination)) {
+      return Optional.of(cache.get(metadataId + source + destination));
+    }
     String cqlFilter =
         String.format(
-            "'id' = '%s' AND 'source' = '%s' AND 'destination' = '%s'",
+            "'mcard-id' = '%s' AND 'source' = '%s' AND 'destination' = '%s'",
```
Comment (shaundmorris) [inline replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/ReplicationItemManagerImpl.java]: What is the distinction here between `mcard-id` and `metadataId`?

File: replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/Syncer.java [hunk 3]
```diff
@@ -136,16 +137,17 @@ SyncResponse sync() {
...
+        Optional<ReplicationItem> replicationItem = getReplicationItem(metadata.getId());
         try {
           if (metadata.isDeleted() && replicationItem.isPresent()) {
             doDelete(metadata, replicationItem.get());
-          } else if (replicationItem.isPresent() && destination.exists(metadata)) {
-            doUpdate(metadata, replicationItem.get());
           } else {
-            doCreate(metadata, replicationItem);
+            Metadata remoteMetadata = destination.exists(metadata);
+            if (replicationItem.isPresent() && remoteMetadata != null) {
```
Comment (shaundmorris) [inline replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/Syncer.java:146]: One thought for the future: a replication log or something like that so we can go back and look at the behavior.

File: replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/Syncer.java [hunk 4]
```diff
@@ -185,6 +187,45 @@ SyncResponse sync() {
...
+    Optional<ReplicationItem> getReplicationItem(String metadataId) {
+
+      ReplicationItem item =
+          replicationItemManager.getItem(metadataId, sourceName, destinationName).orElse(null);
+      ReplicationItem reverseDir =
+          replicationItemManager.getItem(metadataId, destinationName, sourceName).orElse(null);
+      if (item == null && reverseDir == null) {
+        return Optional.empty();
+      }
+      if (item != null
```
Comment (shaundmorris) [inline replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/Syncer.java:199]: wouldn't worry about it for now, but seems like you could check for `item != null` for the rest of this block and not check for it multiple times.

File: replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/Syncer.java [hunk 5]
```diff
@@ -336,13 +382,30 @@ private boolean hasResource(Metadata metadata) {
 
     private void addTagsAndLineage(Metadata metadata) {
       metadata.addLineage(sourceName);
+      // if we are copying to a destination that is in the linage then this is a remote update
```
Comment (shaundmorris) [inline replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/Syncer.java:385]: Nice comment. 👍

File: replication-groomer/src/main/java/com/octo/replication/groomer/ReplicationMetacardGroomerPlugin.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+import java.util.Map.Entry;
+import org.apache.shiro.subject.Subject;
+import org.codice.ddf.configuration.SystemInfo;
+import org.codice.ddf.platform.util.uuidgenerator.UuidGenerator;
+import org.codice.ditto.replication.api.Replication;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Copied of the StandardMetacardGroomer and updated for replication purposes. This groomer replaces
```
Comment (shaundmorris) [inline replication-groomer/src/main/java/com/octo/replication/groomer/ReplicationMetacardGroomerPlugin.java]: `Copy` of

File: replication-groomer/src/main/java/com/octo/replication/groomer/ReplicationMetacardGroomerPlugin.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+ * the StandardMetacardGroomer and should be the only groomer running in the system.
```
Comment (shaundmorris) [inline replication-groomer/src/main/java/com/octo/replication/groomer/ReplicationMetacardGroomerPlugin.java:46]: This is a good comment. Not sure if it makes sense to have documented somewhere else eventually so people know to not have the StandardMetacardGroomer running.

**Review by shaundmorris — APPROVED** [review]
I haven't spent much time in the replication code, but looks pretty good from my view. I'm sure I could ask more questions than have more meaningful comments :-)

File: adapters/ddf-adapter/src/main/java/com/connexta/replication/adapters/ddf/csw/CswMinimalRecordConverter.java (hunk 1 above)
Comment (clockard) [reply adapters/ddf-adapter/src/main/java/com/connexta/replication/adapters/ddf/csw/CswMinimalRecordConverter.java:87]: It shouldn't be null but it wouldn't hurt to add a check.

File: replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/ReplicationItemManagerImpl.java (hunk 2 above)
Comment (clockard) [reply replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/ReplicationItemManagerImpl.java]: They are the same thing here.  Poor naming choice on my part. I will fix it.

File: replication-groomer/src/main/java/com/octo/replication/groomer/ReplicationMetacardGroomerPlugin.java (hunk 6 above)
Comment (clockard) [reply replication-groomer/src/main/java/com/octo/replication/groomer/ReplicationMetacardGroomerPlugin.java]: Fixed

File: replication-groomer/src/main/java/com/octo/replication/groomer/ReplicationMetacardGroomerPlugin.java (hunk 7 above)
Comment (clockard) [reply replication-groomer/src/main/java/com/octo/replication/groomer/ReplicationMetacardGroomerPlugin.java:46]: Agree. Not sure where we are documenting things like that

**Comment by clockard** [comment]
build now

**Comment by clockard** [comment]
build now

**Comment by clockard** [comment]
build now

**Comment by clockard** [comment]
build now

**Pull request merged**